### ISSUE TITLE
RFE:  Princess needs "Hazard" terrain list updated

### ISSUE THREAD

**Issue by Thom293** [MegaMek/megamek#5361]
Im putting this up to remind myself, but its for anyone that wants to handle it.  I thought I had put it up before but I guess not.

The below would need similar fixes to the Ice fix I did before.  I think.

Right now, princess does not see at least the following hexes as a hazard, or doesnt see them as a hazard in all of the necessary situations:

Swamp
Mud
Quicksand

She will gleefully walk/run/jump into any of those hex and strand herself even when there are ample other alternatives.  There are probably more but I dont have my notes as Im not at home.

On the opposite end of the spectrum, she will not move underwater units, or at least some of them, at all.  My guess is that she sees high depth water hexes like a land unit would and so doesnt move UWU into them.  I think a UWU move exception needs to be coded for water, and that would solve the issue.  That is just a guess though.  Related:  https://github.com/MegaMek/megamek/issues/4902

**Comment by repligator** [comment]
I need to look into this more but I don't think Princess understands black ice either.

**Comment by Thom293** [comment]
Lave and related hotfoot things.  May have a few other things to add too.

**Comment by Thom293** [comment]
It was suggested a few weeks back that I ping some old bugs before .50 ships, just in case people are looking. I forgot to do so, so am doing now.

This is an old persistent one that is related to MM basic map navigation and it would be nice if it were fixed, if someone has time to look at it. Thank you.

Related to #5614

**Comment by Sleet01** [comment]
I can take a short pass at this today.
We will want to _brutally_ test any changes prior to cutting the next release, however.

**Comment by Thom293** [comment]
If you are able to get these working I will commit to testing them.  Thank you.

**Comment by Sleet01** [comment]
I will examine the following:

Swamp
Mud
Quicksand
Magma 
UWU underwater

Lava doesn't seem to be defined separately; it is just a synonym for "Magma" used in some of the map-tagging code.

**Comment by Thom293** [comment]
I'm not super sure on lava. One is open firey stuff and one is firey stuff under a crust you can fall through. So akin to water under ice, I'd guess. I don't have access to my books just now. Sorry.

**Comment by Sleet01** [comment]
Yeah, we model that as "Crust" and "Liquid" Magma, which is how TO:AR defines it.

Looks like "Lava" is a special form of flooding Magma and will require flow modeling, so that should get wrapped into future Flood work.

**Comment by Thom293** [comment]
Also I'm not certain on the UMU.  If you would like testing assistance while going through this let me know. I'm available in about 6 hours.

**Comment by Sleet01** [comment]
I think you're right about UMU: they're not being taken into account for several unit types.  Also SUBMARINE movement should grant immunity.

**Comment by IllianiCBT** [comment]
The frequent use of UwU in this thread is very amusing to read

**Comment by IllianiCBT** [comment]
Wait, you guys are saying UMU - pardon my dyslexia, uwu xD

**Comment by Sleet01** [comment]
> Wait, you guys are saying UMU - pardon my dyslexia, uwu xD

There might have been, uh, some secret editing on my part >_>

**Comment by Thom293** [comment]
To be fair, I'm not confident on what either means and have seen both used.  I'm just using the last one someone else used to hide my ignorance.

**Comment by repligator** [comment]
A couple of things to add:
Tundra - Chance to bog down. A slightly less punishing Deep Snow/Mud.
Black Ice - I wrote (read: copied and pasted from Ice) the code that keeps Princess from stepping on known Black Ice. I am absolutely clueless how to quantify or convey how potential Black Ice hexes should be avoided.
"Dear Princess, Please avoid hexes that might have Black Ice, which for some maps, is the majority of the hexes, which means you can't really avoid them. Okay? Bye."

**Comment by Sleet01** [comment]
> A couple of things to add: Tundra - Chance to bog down. A slightly less punishing Deep Snow/Mud. Black Ice - I wrote (read: copied and pasted from Ice) the code that keeps Princess from stepping on known Black Ice. I am absolutely clueless how to quantify or convey how potential Black Ice hexes should be avoided. "Dear Princess, Please avoid hexes that might have Black Ice, which for some maps, is the majority of the hexes, which means you can't really avoid them. Okay? Bye."

The standard approach, and I think it would work here, is to multiply the hazard factor of a _definite_ black ice hex by the chance that any given hex _is_ black ice.  This is how the Magma Crust and Liquid hazards are adjusted relative to each other: the chance of Magma Crust turning into Magma Liquid is 1/6 if walking, 3/6 if jumping onto it.  But the expected damage from a given hex of Magma Liquid is the same either way, so you can quantify the relative hazard by multiplying by the chance of having the Magma Crust become Magma Liquid.

**Comment by Sleet01** [comment]
> I'm not super sure on lava. One is open firey stuff and one is firey stuff under a crust you can fall through. So akin to water under ice, I'd guess. I don't have access to my books just now. Sorry.

I think I've found why Princess wasn't considering Liquid Magma very scary: she never considered units could get stuck in it!  Or that hover units _that take damage_ might drop into it.
I'll try to fix that up.

**Comment by pheonixstorm** [comment]
> Black Ice - I wrote (read: copied and pasted from Ice) the code that keeps Princess from stepping on known Black Ice. I am absolutely clueless how to quantify or convey how potential Black Ice hexes should be avoided. "Dear Princess, Please avoid hexes that might have Black Ice, which for some maps, is the majority of the hexes, which means you can't really avoid them. Okay? Bye."

Blame CGL for those rules, I just wrote the code RAW. Or at least best I could. If using that option just use careful movement.

Which could be something Princess could be set to do. If Black Ice then use careful movement on pavement/road. Not an AI guy so that code is far far out of my wheelhouse.

**Comment by Thom293** [comment]
Here is a thing I should bring up just in case.  Maybe you already know.  

When I did the ice/psr fix I basically completely forbade those hexes.  Which works ok.  But once combat is entered she appears to throw out all caution whatsoever and will go back to craziness.   So, there might be two behavior states or some weight that turns on that overrides whatever you do.  That said, even just avoiding out of combat would be a huge improvement.

**Comment by Sleet01** [comment]
> Here is a thing I should bring up just in case. Maybe you already know.
> 
> When I did the ice/psr fix I basically completely forbade those hexes. Which works ok. But once combat is entered she appears to throw out all caution whatsoever and will go back to craziness. So, there might be two behavior states or some weight that turns on that overrides whatever you do. That said, even just avoiding out of combat would be a huge improvement.

I think the issue is that the hazard value is only half of the equation.  Things like expected incoming damage, the ability to kill an enemy target, etc. all get balanced against each other.  A hazardous hex may end up still having the highest (even if negative) value overall.

**Comment by pheonixstorm** [comment]
> > I'm not super sure on lava. One is open firey stuff and one is firey stuff under a crust you can fall through. So akin to water under ice, I'd guess. I don't have access to my books just now. Sorry.
> 
> I think I've found why Princess wasn't considering Liquid Magma very scary: she never considered units could get stuck in it! Or that hover units _that take damage_ might drop into it. I'll try to fix that up.

@Sleet01 
RAW mechs should not be getting stuck unless they jump into the stuff, even then though you should avoid magma like the plague regardless due to the heat and damage it causes.

**Comment by Sleet01** [comment]
> > > I'm not super sure on lava. One is open firey stuff and one is firey stuff under a crust you can fall through. So akin to water under ice, I'd guess. I don't have access to my books just now. Sorry.
> > 
> > 
> > I think I've found why Princess wasn't considering Liquid Magma very scary: she never considered units could get stuck in it! Or that hover units _that take damage_ might drop into it. I'll try to fix that up.
> 
> @Sleet01 RAW mechs should not be getting stuck unless they jump into the stuff, even then though you should avoid magma like the plague regardless due to the heat and damage it causes.

Right, but none of that is currently being considered.

**Comment by Thom293** [comment]
I tend to oversimplify but reducing the chance of zero or near zero to walk on lava magma mud quicksand etc where there is literally any other option seems to make most sense.

**Comment by pheonixstorm** [comment]
My guess would be to find the behavioral code that the AI sliders use. That could be a starting point to help figure out what factors goes into the decision tree. Don't know if searching for that PR (if there was one) or just diving into the code would be quicker.

**Comment by Sleet01** [comment]
> My guess would be to find the behavioral code that the AI sliders use. That could be a starting point to help figure out what factors goes into the decision tree. Don't know if searching for that PR (if there was one) or just diving into the code would be quicker.

Just to be clear, this RFE work will _only_ cover adding some terrain to the existing hazard analysis and will not include any behavior modifications.

**Comment by Thom293** [comment]
K Sleet.  Let me know when/what you need testing help with.  No rush.

**Comment by Sleet01** [comment]
Got some good candidates done; I'll write up some unit tests and get them up for PR tomorrow morning.  Goal is getting code into the nightly tomorrow, or at least a bit more tested.

BTW, it looks like any changes to prevent deployment issues would go in `megamek.client.bot.BotClient#getStartingCoordsArray()`, which averages between 10 and 20 year-old code.
Currently it doesn't use any of the hazard detection code from `BasicPathRanker.java`.

**Comment by Thom293** [comment]
Legend!

**Comment by Sleet01** [comment]
@repligator I'll take a look at Tundra and Black Ice if I have time today.

**Comment by Sleet01** [comment]
Closing this as all work is done and merged.  Ended up adding Tundra, Black Ice, and Deep Snow on top of original terrain stuff.
